File: Music.Android/short/music/src/main/java/com/zitherharp/music/core/Youtube.java
package com.zitherharp.music.core;

import com.zitherharp.music.model.Artist;
import com.zitherharp.music.type.Language;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public abstract class Youtube extends Spreadsheet {
    protected String artistId;
    protected int duration = 0;
    protected List<Artist> artists;

    public enum Image {
        DEFAULT("120x90"),
        MQDEFAULT("320x180"),
        HQDEFAULT("480x360"),
        SDDEFAULT("640x480"),
        MAXRESDEFAULT("1280x720");
        String size;
        Image(String size) {
            this.size = size;
        }
        public String getSize() {
            return size;
        }
    }

    public List<Artist> getArtists() {
        if (artists == null) {
            artists = new ArrayList<>();
            for (String id : artistId.split(splitChar)) {
                for (Artist artist : Repository.getArtists()) {
                    if (artist.Id.equals(id)) {
                        artists.add(artist); break;
                    }
                }
            }
        }
        return artists;
    }

    public String getImageUrl(@NotNull Image image) {
        return String.format("https://i.ytimg.com/vi/%s/%s.jpg", Id, image.name().toLowerCase());
    }

    public String getShareUrl(boolean isEmbed) {
        return !isEmbed ? "https://youtu.be/" + Id : "https://www.youtube.com/embed/" + Id;
    }

    public String toString(Language language) {
        return getName(language) + " - ";
    }
}

File: Music.Client.Android/Kotlin/Music/lib/src/main/java/com/zitherharp/music/core/Youtube.kt
package com.zitherharp.music.core

abstract class Youtube(id: String): Spreadsheet(id) {
    var duration = 0
    var viewCount = 0
    var commentCount = 0

    companion object {
        const val DURATION = 4
    }

    enum class Image {
        DEFAULT,
        MQDEFAULT,
        HQDEFAULT,
        SDDEFAULT,
        MAXRESDEFAULT
    }

    fun getImageUrl(image: Image) =
        "https://i.ytimg.com/vi/$id/${image.name.lowercase()}.jpg"

    fun getShareUrl(isEmbed: Boolean) =
        if (!isEmbed) "https://youtu.be/$id" else "https://www.youtube.com/embed/$id"
}
